src/storage.rs

```diff
@@ -126,136 +172,164 @@ impl MemStorageCore {
+        if compact_index <= self.inner_first_index() {
+            // Don't need to treat this case as an error.
+            return Ok(());
         }
-        if compact_index > self.inner_last_index() {
+
+        if compact_index > self.inner_last_index() + 1 {
             panic!(
-                "compact {} is out of bound lastindex({})",
+                "compact not received raft logs: {}, last index: {}",
                 compact_index,
                 self.inner_last_index()
-            )
+            );
         }
 
-        let i = (compact_index - offset) as usize;
-        let entries = self.entries.drain(i..).collect();
-        self.entries = entries;
+        if let Some(entry) = self.entries.first() {
+            let offset = compact_index - entry.get_index();
+            self.entries.drain(..offset as usize);
+        }
         Ok(())
     }
 
     /// Append the new entries to storage.
-    /// TODO: ensure the entries are continuous and
-    /// entries[0].get_index() > self.entries[0].get_index()
     pub fn append(&mut self, ents: &[Entry]) -> Result<()> {
         if ents.is_empty() {
             return Ok(());
         }
-        let first = self.entries[0].get_index() + 1;
-        let last = ents[0].get_index() + ents.len() as u64 - 1;
-
-        if last < first {
-            return Ok(());
-        }
-        // truncate compacted entries
-        let te: &[Entry] = if first > ents[0].get_index() {
-            let start_ent = (first - ents[0].get_index()) as usize;
-            &ents[start_ent..]
-        } else {
-            ents
-        };
-
-        let offset = te[0].get_index() - self.entries[0].get_index();
-        if self.entries.len() as u64 > offset {
-            let mut new_entries: Vec<Entry> = vec![];
-            new_entries.extend_from_slice(&self.entries[..offset as usize]);
-            new_entries.extend_from_slice(te);
-            self.entries = new_entries;
-        } else if self.entries.len() as u64 == offset {
-            self.entries.extend_from_slice(te);
-        } else {
+        if self.inner_first_index() > ents[0].get_index() {
             panic!(
-                "missing log entry [last: {}, append at: {}]",
-                self.inner_last_index(),
-                te[0].get_index()
-            )
+                "overwrite compacted raft logs, compacted: {}, append: {}",
+                self.inner_first_index() - 1,
+                ents[0].get_index(),
+            );
         }
+        if self.inner_last_index() + 1 < ents[0].get_index() {
+            // Gaps are not allowed in raft logs.
+            return Err(Error::Store(StorageError::LogGap));
+        }
+
+        // Remove all entries overwritten by `ents`, and truncate `ents` if need.
+        let diff = ents[0].get_index() - self.inner_first_index();
+        self.entries.drain(diff as usize..);
+        ents.iter().for_each(|e| self.entries.push(e.clone()));
```

Comment:
Maybe you can use `self.entries.append(ents)`?

Reply:
I don't think so (`ents` isn't mutable) but I think you can use `self.entries.extend_from_slice(ents)`